class-bookrestcontroller.php

```diff
@@ -0,0 +1,307 @@
+			'id'      => $next_id + 1,
+			'title'   => $request->get_param('title'),
+			'author'  => $request->get_param('author'),
+			'genre'   => $request->get_param('genre'),
+			'summary' => $request->get_param('summary'),
+		);
+		if ( $data === false ) {
+			$data = array();
+		}
+		$data[ $model['id'] ] = $model;
+		if ( ! update_post_meta( $request->get_param('post_id'), self::META_KEY, $data ) ) {
+			$response->set_status(400);
+		}
+		$response->set_data($this->escape_html($model));
+		return $response;
+	}
+	/**
+	 * Updates only the specified fields of a book from the serialized array
+	 *
+	 * @param WP_REST_Request $request MUST have at least two keys, 'id' and 'post_id'
+	 *
+	 * @return WP_REST_Response Whether the operation was successful or not as JSON
+	 */
+	public function update_item( $request ) {
+		$response = new WP_REST_Response();
+		$id       = (int) $request->get_param('id');
+		$post_id  = (int) $request->get_param('post_id');
+		$data     = get_post_meta( $post_id, self::META_KEY, true );
+		// inner model is used for cases where request doesn't have all fields
+		$inner_model = $data[ $id ];
+		$model       = array(
+			'id'      => $id,
+			'title'   => $request->get_param( 'title' ) !== null ? $request->get_param('title') : $inner_model['title'],
+			'author'  => $request->get_param( 'author' ) !== null ? $request->get_param('author') : $inner_model['author'],
+			'genre'   => $request->get_param( 'genre' ) !== null ? $request->get_param('genre') : $inner_model['genre'],
+			'summary' => $request->get_param( 'summary' ) !== null ? $request->get_param('summary') : $inner_model['summary'],
+		);
+		//now add model into $data, serialize data then update the post meta
+		$data[ $id ] = $model;
+		if ( ! update_post_meta( $post_id, self::META_KEY, $data ) ) {
+			$response->set_status(400);
+		}
+		$response->set_data($this->escape_html($model));
+		return $response;
+	}
+	/**
+	 * Deletes one item from the serialized array
+	 *
+	 * @param WP_REST_Request $request  MUST have two keys 'id' and 'post_id'
+	 *
+	 * @return WP_REST_Response Whether the operation was successful as JSON
+	 */
+	public function delete_item( $request ) {
+		$response = new WP_REST_Response();
+		$data     = get_post_meta( (int) $request->get_param('post_id'), self::META_KEY, true );
+		unset( $data[ (int) $request['id'] ] );
+		if ( ! update_post_meta( (int) $request->get_param('post_id'), self::META_KEY, $data ) ) {
+			$response->set_status(400);
+		}
+		return $response;
+	}
+	/** Getter for user_can_edit property
+	 * @return bool Whether the user can edit
+	 */
+	public function get_user_can_edit() {
+		return $this->user_can_edit;
+	}
+	/**
+	 * Check if the given parameter is a valid integer
+	 *
+	 * @param mixed $value the value to be checked
+	 *
+	 * @return bool Whether the input is a valid integer
+	 */
+	public function validate_int( $value ) {
+		return isset( $value ) && is_numeric( $value ) && ( (int) $value !== 0 );
```

Comment:
```suggestion
		return ! empty( $value ) && is_numeric( $value );
```

Please look at the difference between isset and empty

Reply:
yep, understood